subprojects/functional/src/main/java/org/gradle/internal/io/IoSupplier.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.io;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.function.Supplier;
+
+/**
+ * A variant of {@link Supplier} that is allowed to throw {@link IOException}.
+ */
+@FunctionalInterface
+public interface IoSupplier<T> {
```

Comment:
I would have preferred to add the functional interfaces which deal with IOExceptions in a separate PR, where we also would use those abstractions somewhere. Currently, those classes are not used.